sls_api/endpoints/tools/files.py

```diff
@@ -443,6 +448,216 @@ def get_file_tree(project, file_path=None):
+      HTTP status code. The JSON response has the following structure:
+
+        {
+            "success": bool,
+            "message": str,
+            "data": object or null
+        }
+
+      - `success`: A boolean indicating whether the operation was successful.
+      - `message`: A string containing a descriptive message about the result.
+      - `data`: On success, an object containing the extracted metadata;
+        `null` on error.
+
+    Example Request:
+
+        GET /my_project/get_metadata_from_xml/by_path/documents/file.xml
+
+    Example Success Response (HTTP 200):
+
+        {
+            "success": true,
+            "message": "Metadata retrieved from XML file.",
+            "data": {
+                "name": "Publication Title",
+                "original_publication_date": "1854-07-20",
+                "language": "en",
+                "genre": "prose"
+            }
+        }
+
+    Example Error Response (HTTP 404):
+
+        {
+            "success": false,
+            "message": "Error: the requested file was not found in the git repository.",
+            "data": null
+        }
+
+    Status Codes:
+
+    - 200 - OK: The request was successful, and the metadata is returned.
+    - 400 - Bad Request: Invalid request parameters (e.g., invalid file
+            path, missing .xml extension, file size exceeds limit).
+    - 403 - Forbidden: Permission denied when trying to read the XML file.
+    - 404 - Not Found: The requested file does not exist.
+    - 500 - Internal Server Error: An unexpected error occurred on the server.
+    """
+    # Validate project config
+    config = get_project_config(project)
+    if config is None:
+        return create_error_response("Error: project config not found on the server.", 500)
+
+    config_ok = check_project_config(project)
+    if not config_ok[0]:
+        return create_error_response(f"Error: {config_ok[1]}", 500)
+
+    # Safely join the base directory and file path
+    full_path = safe_join(config["file_root"], file_path)
+    if full_path is None:
+        return create_error_response("Error: invalid file path.", 400)
+
+    # Resolve the real, absolute paths
+    base_dir = os.path.realpath(config["file_root"])
+    full_path = os.path.realpath(full_path)
+
+    # Verify that full_path is within base_dir, i.e. file_root specified
+    # in config
+    if os.path.commonpath([base_dir, full_path]) != base_dir:
+        return create_error_response("Error: invalid file path.", 400)
+
+    # Check if the file exists
+    try:
+        if not os.path.isfile(full_path):
+            return create_error_response("Error: the requested file was not found on the server.", 404)
+    except Exception:
+        logger.exception(f"Error accessing file at {full_path}")
+        return create_error_response(f"Error accessing file at {file_path}", 500)
+
+    # Check that the file has a .xml extension
+    if not full_path.endswith(".xml"):
```

Comment:
ideally, use `os.path.splitext` to get the extension instead. Using `endswith()` can result in unexpected behavior with some filenames.